tests/core/pyspec/eth2spec/test/helpers/fork_choice.py

```diff
@@ -73,25 +105,53 @@ def on_tick_and_append_step(spec, store, time, test_steps):
     test_steps.append({'tick': int(time)})
 
 
-def run_on_block(spec, store, signed_block, test_steps, valid=True):
+def run_on_block(spec, store, signed_block, valid=True):
     if not valid:
         try:
             spec.on_block(store, signed_block)
-
         except AssertionError:
             return
         else:
             assert False
 
     spec.on_block(store, signed_block)
+    assert store.blocks[signed_block.message.hash_tree_root()] == signed_block.message
+
+
+def add_block(spec, store, signed_block, test_steps, valid=True, allow_invalid_attestations=False):
+    """
+    Run on_block and on_attestation
+    """
     yield get_block_file_name(signed_block), signed_block
+
+    if not valid:
+        try:
+            run_on_block(spec, store, signed_block, valid=True)
+        except AssertionError:
+            test_steps.append({
+                'block': get_block_file_name(signed_block),
+                'valid': False,
+            })
+            return
+        else:
+            assert False
+
+    run_on_block(spec, store, signed_block, valid=True)
     test_steps.append({'block': get_block_file_name(signed_block)})
 
     # An on_block step implies receiving block's attestations
-    for attestation in signed_block.message.body.attestations:
-        spec.on_attestation(store, attestation)
+    try:
+        for attestation in signed_block.message.body.attestations:
+            run_on_attestation(spec, store, attestation, valid=True)
+    except AssertionError:
+        if allow_invalid_attestations:
+            pass
+        else:
+            raise
```

Comment:
It is possible that the `on_block` call successes but the `run_on_attestation` call fails. I added a workaround flag for now until we figure it out.